example/sound.cpp

```diff
@@ -4,25 +4,98 @@
 #include <dr_libs/dr_wav.h>
 
 #include <iostream>
+#include <type_traits>
 #include <vector>
 
+namespace {
+static constexpr int fail_code = 2;
+
+[[noreturn]] void on_error(ALenum error, int line) {
+	std::cerr << "Misc OpenAL error: " << std::hex << error << " on line " << std::dec << line << std::endl;
+	std::exit(fail_code);
+}
+
+void al_check(int line = __builtin_LINE()) {
+	if (auto err = alGetError(); err != AL_NO_ERROR) { on_error(err, line); }
+}
+
+class al_instance final {
+  public:
+	al_instance() noexcept : m_device(alcOpenDevice(nullptr)) {
+		if (!m_device) {
+			std::cerr << "Couldn't initialize OpenAL device" << std::endl;
+		} else {
+			m_context = alcCreateContext(m_device, nullptr);
+			alcMakeContextCurrent(m_context);
+		}
+	}
+	al_instance(al_instance&& rhs) noexcept { exchg(*this, rhs); }
+	al_instance& operator=(al_instance rhs) noexcept { return (exchg(*this, rhs), *this); }
+	~al_instance() {
+		if (valid()) {
+			alcMakeContextCurrent(nullptr);
+			alcDestroyContext(m_context);
+			alcCloseDevice(m_device);
+		}
+	}
+	static void exchg(al_instance& lhs, al_instance& rhs) noexcept {
+		std::swap(lhs.m_device, rhs.m_device);
+		std::swap(lhs.m_context, rhs.m_context);
+	}
+
+	bool valid() const noexcept { return m_device && m_context; }
+	explicit operator bool() const noexcept { return valid(); }
```

Comment:
I'd stay away from operators like these even if they are explicit, `if(instance.valid())` is much clearer than `if(instance)` (maybe second one could mean instance exists but is not valid i.e. usable)

Reply:
What's your reasoning? Explicit boolean operators are quite common throughout stdlib too, eg `ifstream`. Besides, the function also exists.

Reply:
i'm maybe too used to rust